src/Components/WebAssembly/JSInterop/src/WebAssemblyJSRuntime.cs

```diff
@@ -60,32 +61,52 @@ protected override void EndInvokeDotNet(DotNetInvocationInfo callInfo, in DotNet
             BeginInvokeJS(0, "DotNet.jsCallDispatcher.endInvokeDotNetFromJS", args, JSCallResultType.Default, 0);
         }
 
-        /// <inheritdoc />
-        TResult IJSUnmarshalledRuntime.InvokeUnmarshalled<TResult>(string identifier)
-            => ((IJSUnmarshalledRuntime)this).InvokeUnmarshalled<object, object, object, TResult>(identifier, null, null, null);
-
-        /// <inheritdoc />
-        TResult IJSUnmarshalledRuntime.InvokeUnmarshalled<T0, TResult>(string identifier, T0 arg0)
-            => ((IJSUnmarshalledRuntime)this).InvokeUnmarshalled<T0, object, object, TResult>(identifier, arg0, null, null);
-
-        /// <inheritdoc />
-        TResult IJSUnmarshalledRuntime.InvokeUnmarshalled<T0, T1, TResult>(string identifier, T0 arg0, T1 arg1)
-            => ((IJSUnmarshalledRuntime)this).InvokeUnmarshalled<T0, T1, object, TResult>(identifier, arg0, arg1, null);
-
-        /// <inheritdoc />
-        TResult IJSUnmarshalledRuntime.InvokeUnmarshalled<T0, T1, T2, TResult>(string identifier, T0 arg0, T1 arg1, T2 arg2)
+        internal TResult InvokeUnmarshalled<T0, T1, T2, TResult>(string identifier, T0 arg0, T1 arg1, T2 arg2, long targetInstanceId)
         {
+            var resultType = JSCallResultTypeHelper.FromGeneric<TResult>();
+
             var callInfo = new JSCallInfo
             {
                 FunctionIdentifier = identifier,
-                ResultType = JSCallResultTypeHelper.FromGeneric<TResult>()
+                TargetInstanceId = targetInstanceId,
+                ResultType = resultType,
             };
 
-            var result = InternalCalls.InvokeJS<T0, T1, T2, TResult>(out var exception, ref callInfo, arg0, arg1, arg2);
+            string exception;
+            TResult result;
+
+            switch (resultType)
+            {
+                case JSCallResultType.Default:
+                    result = InternalCalls.InvokeJS<T0, T1, T2, TResult>(out exception, ref callInfo, arg0, arg1, arg2);
+                    break;
+                case JSCallResultType.JSObjectReference:
+                    var id = InternalCalls.InvokeJS<T0, T1, T2, int>(out exception, ref callInfo, arg0, arg1, arg2);
+                    result = (TResult)(object)new WebAssemblyJSObjectReference(this, id);
```

Comment:
Is there any possible drawback to executing this line (85) when `exception` is nonnull?

I know we don't necessarily have to optimize every last drop when it comes to calls that are going to throw (since we don't consider exception-throwing to be part of any reasonably-optimized call flow) but it's potentially confusing that we're instantiating an `IDisposable` instance that we know will never get disposed, because we're not returning it to the caller. To avoid confusion it might be preferable to duplicate lines 91-93 in each of the two cases of this switch before instantiating the `WebAssemblyJSObjectReference`.